File: src/extras-kcoreaddons/libkurlmimedata.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const QMimeData = @import("libqt6").QMimeData;
const QUrl = @import("libqt6").QUrl;
const kurlmimedata_enums = enums;
const std = @import("std");
const ArrayMap_constu8_constu8 = std.array_hash_map.String([]const u8);

/// ### [Upstream resources](https://api.kde.org/kurlmimedata.html)
pub const KUrlMimeData = extern struct {
    /// ### [Upstream resources](https://api.kde.org/kurlmimedata.html)
    ///
    /// The pointer to the underlying Qt C++ object
    ///
    ptr: QtC.KUrlMimeData,

    pub const _is_KUrlMimeData = {};

    /// ### [Upstream resources](https://api.kde.org/kurlmimedata.html#setUrls)
    ///
    /// ## Parameter(s):
    ///
    /// ` urls: []QUrl `
    ///
    /// ` mostLocalUrls: []QUrl `
    ///
    /// ` mimeData: QMimeData `
    ///
    pub fn SetUrls(urls: []QUrl, mostLocalUrls: []QUrl, mimeData: anytype) void {
        const urls_list = qtc.libqt_list{
            .len = urls.len,
            .data = @ptrCast(urls.ptr),
        };
        const mostLocalUrls_list = qtc.libqt_list{
            .len = mostLocalUrls.len,
            .data = @ptrCast(mostLocalUrls.ptr),
        };
        comptime _ = @TypeOf(mimeData)._is_QMimeData;
        qtc.KUrlMimeData_SetUrls(urls_list, mostLocalUrls_list, @ptrCast(mimeData.ptr));
    }

    /// ### [Upstream resources](https://api.kde.org/kurlmimedata.html#exportUrlsToPortal)
    ///
    /// ## Parameter(s):
    ///
    /// ` mimeData: QMimeData `
    ///
    pub fn ExportUrlsToPortal(mimeData: anytype) bool {
        comptime _ = @TypeOf(mimeData)._is_QMimeData;
        return qtc.KUrlMimeData_ExportUrlsToPortal(@ptrCast(mimeData.ptr));
    }

    /// ### [Upstream resources](https://api.kde.org/kurlmimedata.html#setMetaData)
    ///
    /// ## Parameter(s):
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ` metaData: ArrayMap_constu8_constu8 `
    ///
    /// ` mimeData: QMimeData `
    ///
    pub fn SetMetaData(allocator: std.mem.Allocator, metaData: ArrayMap_constu8_constu8, mimeData: anytype) void {
        const metaData_count = metaData.count();
        const metaData_keys = allocator.alloc(qtc.libqt_string, metaData_count) catch @panic("kurlmimedata.SetMetaData: Memory allocation failed");
        defer allocator.free(metaData_keys);
        const metaData_values = allocator.alloc(qtc.libqt_string, metaData_count) catch @panic("kurlmimedata.SetMetaData: Memory allocation failed");
        defer allocator.free(metaData_values);
        var i: usize = 0;
        var metaData_it = metaData.iterator();
        while (metaData_it.next()) |it_entry| : (i += 1) {
            const metaData_key = it_entry.key_ptr.*;
            metaData_keys[i] = qtc.libqt_string{
                .len = metaData_key.len,
                .data = metaData_key.ptr,
            };
            const value = it_entry.value_ptr.*;
            metaData_values[i] = qtc.libqt_string{
                .len = value.len,
                .data = value.ptr,
            };
        }
        const metaData_map = qtc.libqt_map{
            .len = metaData_count,
            .keys = @ptrCast(metaData_keys.ptr),
            .values = @ptrCast(metaData_values.ptr),
        };
        comptime _ = @TypeOf(mimeData)._is_QMimeData;
        qtc.KUrlMimeData_SetMetaData(metaData_map, @ptrCast(mimeData.ptr));
    }

    /// ### [Upstream resources](https://api.kde.org/kurlmimedata.html#mimeDataTypes)
    ///
    /// ## Parameter(s):
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn MimeDataTypes(allocator: std.mem.Allocator) []const []const u8 {
        const _arr: qtc.libqt_list = qtc.KUrlMimeData_MimeDataTypes();
        var _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i|
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]const u8, _arr.len) catch @panic("kurlmimedata.MimeDataTypes: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("kurlmimedata.MimeDataTypes: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// ### [Upstream resources](https://api.kde.org/kurlmimedata.html#urlsFromMimeData)
    ///
    /// ## Parameter(s):
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ` mimeData: QMimeData `
    ///
    /// ` decodeOptions: flag of kurlmimedata_enums.DecodeOption `
    ///
    /// ` metaData: ArrayMap_constu8_constu8 `
    ///
    pub fn UrlsFromMimeData(allocator: std.mem.Allocator, mimeData: anytype, decodeOptions: i32, metaData: ArrayMap_constu8_constu8) []QUrl {
        comptime _ = @TypeOf(mimeData)._is_QMimeData;
        const metaData_count = metaData.count();
        const metaData_keys = allocator.alloc(qtc.libqt_string, metaData_count) catch @panic("kurlmimedata.UrlsFromMimeData: Memory allocation failed");
        defer allocator.free(metaData_keys);
        const metaData_values = allocator.alloc(qtc.libqt_string, metaData_count) catch @panic("kurlmimedata.UrlsFromMimeData: Memory allocation failed");
        defer allocator.free(metaData_values);
        var i: usize = 0;
        var metaData_it = metaData.iterator();
        while (metaData_it.next()) |it_entry| : (i += 1) {
            const metaData_key = it_entry.key_ptr.*;
            metaData_keys[i] = qtc.libqt_string{
                .len = metaData_key.len,
                .data = metaData_key.ptr,
            };
            const value = it_entry.value_ptr.*;
            metaData_values[i] = qtc.libqt_string{
                .len = value.len,
                .data = value.ptr,
            };
        }
        var metaData_map = qtc.libqt_map{
            .len = metaData_count,
            .keys = @ptrCast(metaData_keys.ptr),
            .values = @ptrCast(metaData_values.ptr),
        };
        const _arr: qtc.libqt_list = qtc.KUrlMimeData_UrlsFromMimeData(@ptrCast(mimeData.ptr), @bitCast(decodeOptions), &metaData_map);
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QUrl, _arr.len) catch @panic("kurlmimedata.UrlsFromMimeData: Memory allocation failed");
        const _data: [*]QtC.QUrl = @ptrCast(@alignCast(_arr.data));
        for (0.._arr.len) |ii|
            _ret[ii] = .{ .ptr = _data[ii] };
        return _ret;
    }
};

/// ### [Upstream resources](https://api.kde.org/kurlmimedata.html#public-types)
pub const enums = struct {
    pub const DecodeOption = enum(i32) {
        pub const PreferKdeUrls: i32 = 0;
        pub const PreferLocalUrls: i32 = 1;
    };
};

File: src/extras-kcoreaddons/libkurlmimedata.cpp
#include <KUrlMimeData>
#include <QList>
#include <QMap>
#include <QMimeData>
#include <QString>
#include <QUrl>
#include <kurlmimedata.h>
#include "libkurlmimedata.h"
#include "libkurlmimedata.hxx"

void KUrlMimeData_SetUrls(const libqt_list /* of QUrl* */ urls, const libqt_list /* of QUrl* */ mostLocalUrls, QMimeData* mimeData) {
    QList<QUrl> urls_QList;
    urls_QList.reserve(urls.len);
    QUrl** urls_arr = static_cast<QUrl**>(urls.data);
    for (size_t i = 0; i < urls.len; ++i) {
        urls_QList.push_back(*(urls_arr[i]));
    }
    QList<QUrl> mostLocalUrls_QList;
    mostLocalUrls_QList.reserve(mostLocalUrls.len);
    QUrl** mostLocalUrls_arr = static_cast<QUrl**>(mostLocalUrls.data);
    for (size_t i = 0; i < mostLocalUrls.len; ++i) {
        mostLocalUrls_QList.push_back(*(mostLocalUrls_arr[i]));
    }
    KUrlMimeData::setUrls(urls_QList, mostLocalUrls_QList, mimeData);
}

bool KUrlMimeData_ExportUrlsToPortal(QMimeData* mimeData) {
    return KUrlMimeData::exportUrlsToPortal(mimeData);
}

void KUrlMimeData_SetMetaData(const libqt_map /* of libqt_string to libqt_string */ metaData, QMimeData* mimeData) {
    QMap<QString, QString> metaData_QMap;
    libqt_string* metaData_karr = static_cast<libqt_string*>(metaData.keys);
    libqt_string* metaData_varr = static_cast<libqt_string*>(metaData.values);
    for (size_t i = 0; i < metaData.len; ++i) {
        QString metaData_karr_i_QString = QString::fromUtf8(metaData_karr[i].data, metaData_karr[i].len);
        QString metaData_varr_i_QString = QString::fromUtf8(metaData_varr[i].data, metaData_varr[i].len);
        metaData_QMap[metaData_karr_i_QString] = metaData_varr_i_QString;
    }
    KUrlMimeData::setMetaData(metaData_QMap, mimeData);
}

libqt_list /* of libqt_string */ KUrlMimeData_MimeDataTypes() {
    QList<QString> _ret = KUrlMimeData::mimeDataTypes();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size())));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        auto _lv_ret = _ret[i];
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_b = _lv_ret.toUtf8();
        libqt_string _lv_str;
        _lv_str.len = _lv_b.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_b.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_list /* of QUrl* */ KUrlMimeData_UrlsFromMimeData(const QMimeData* mimeData, int decodeOptions, libqt_map* /* of libqt_string to libqt_string */ metaData) {
    QMap<QString, QString>* metaData_QMap;
    libqt_string* metaData_karr = static_cast<libqt_string*>(metaData->keys);
    libqt_string* metaData_varr = static_cast<libqt_string*>(metaData->values);
    for (size_t i = 0; i < metaData->len; ++i) {
        QString metaData_karr_i_QString = QString::fromUtf8(metaData_karr[i].data, metaData_karr[i].len);
        QString metaData_varr_i_QString = QString::fromUtf8(metaData_varr[i].data, metaData_varr[i].len);
        (*metaData_QMap)[metaData_karr_i_QString] = metaData_varr_i_QString;
    }
    QList<QUrl> _ret = KUrlMimeData::urlsFromMimeData(mimeData, static_cast<KUrlMimeData::DecodeOptions>(decodeOptions), metaData_QMap);
    // Convert QList<> from C++ memory to manually-managed C memory
    QUrl** _arr = static_cast<QUrl**>(malloc(sizeof(QUrl*) * (_ret.size())));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new QUrl(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}
